core/chains/evm/txmgr/orm.go

```diff
@@ -943,3 +957,259 @@ func (o *orm) SaveReplacementInProgressAttempt(oldAttempt EthTxAttempt, replacem
+			return errors.Wrap(err, "SaveEthTxAttempt failed to save eth_tx_attempt")
+		}
+		return nil
+	})
+}
+
+// Updates eth tx from unstarted to in_progress and inserts in_progress eth attempt
+func (o *orm) UpdateEthTxUnstartedToInProgress(etx *EthTx, attempt *EthTxAttempt, qopts ...pg.QOpt) error {
+	qq := o.q.WithOpts(qopts...)
+	if etx.Nonce == nil {
+		return errors.New("in_progress transaction must have nonce")
+	}
+	if etx.State != EthTxUnstarted {
+		return errors.Errorf("can only transition to in_progress from unstarted, transaction is currently %s", etx.State)
+	}
+	if attempt.State != EthTxAttemptInProgress {
+		return errors.New("attempt state must be in_progress")
+	}
+	etx.State = EthTxInProgress
+	return qq.Transaction(func(tx pg.Queryer) error {
+		query, args, e := tx.BindNamed(insertIntoEthTxAttemptsQuery, attempt)
+		if e != nil {
+			return errors.Wrap(e, "failed to BindNamed")
+		}
+		err := tx.Get(attempt, query, args...)
+		if err != nil {
+			switch e := err.(type) {
+			case *pgconn.PgError:
+				if e.ConstraintName == "eth_tx_attempts_eth_tx_id_fkey" {
+					return errEthTxRemoved
+				}
+			}
+			return errors.Wrap(err, "UpdateEthTxUnstartedToInProgress failed to create eth_tx_attempt")
+		}
+		err = tx.Get(etx, `UPDATE eth_txes SET nonce=$1, state=$2, broadcast_at=$3, initial_broadcast_at=$4 WHERE id=$5 RETURNING *`, etx.Nonce, etx.State, etx.BroadcastAt, etx.InitialBroadcastAt, etx.ID)
+		return errors.Wrap(err, "UpdateEthTxUnstartedToInProgress failed to update eth_tx")
+	})
+}
+
+// GetEthTxInProgress returns either 0 or 1 transaction that was left in
+// an unfinished state because something went screwy the last time. Most likely
+// the node crashed in the middle of the ProcessUnstartedEthTxs loop.
+// It may or may not have been broadcast to an eth node.
+func (o *orm) GetEthTxInProgress(fromAddress common.Address, qopts ...pg.QOpt) (etx *EthTx, err error) {
+	qq := o.q.WithOpts()
+	etx = new(EthTx)
+	err = qq.Transaction(func(tx pg.Queryer) error {
+		err = qq.Get(etx, `SELECT * FROM eth_txes WHERE from_address = $1 and state = 'in_progress'`, fromAddress.Bytes())
+		if errors.Is(err, sql.ErrNoRows) {
+			return err
+		} else if err != nil {
+			return errors.Wrap(err, "GetEthTxInProgress failed while loading eth tx")
+		}
+		if err = o.LoadEthTxAttempts(etx, pg.WithQueryer(tx)); err != nil {
+			return errors.Wrap(err, "GetEthTxInProgress failed while loading EthTxAttempts")
+		}
+		if len(etx.EthTxAttempts) != 1 || etx.EthTxAttempts[0].State != EthTxAttemptInProgress {
+			return errors.Errorf("invariant violation: expected in_progress transaction %v to have exactly one unsent attempt. "+
+				"Your database is in an inconsistent state and this node will not function correctly until the problem is resolved", etx.ID)
+		}
+		return nil
+	})
+
+	if errors.Is(err, sql.ErrNoRows) {
+		return nil, nil
```

Comment:
This return is inconsistent

Reply:
If there are no rows, then there's no error but the *etx is nil. (taken directly from https://github.com/smartcontractkit/chainlink/pull/8584/files/f6e0eaa3c430c165b5c6e53cb618ce10703521f2#diff-bdf2616ee250413e0d4b2e171adb1819033e3280cb122ec672b51eef64c52c7fL444)

Reply:
I mean, its inconsistent with this error message: https://github.com/smartcontractkit/chainlink/pull/8584/files#diff-17acb6da51a93f6407f839210ff4ecd5809bc34eccc3252aa5e613429d4ca386R81

Either return an error for zero rows or don't. Very confusing to sometimes return error and sometimes return nil.

Reply:
I think it depends on the context of the method. `GetEthTxInProgress` is ok with returning 0 rows (hence nil return) because you might not have any in progress transactions at the moment.

However, on the other hand it's not okay if `UpdateEthKeyNextNonce` affects 0 rows because it should always update the next nonce.

I can rename the `ErrZeroRowsAffected` to be key specific if that makes things less confusing